Calculate rewards from genesis parameters

Anviking opened the pull request:
# Issue Number
#1276 

# Overview

- e2c8683848493634e8bcb13930b9e8e6bb6f5f5b
  [WIP] Calculate rewards from block0 params
  I am trying to keep the reward-calculation jörmungandr-specific. I don't
want to extend BlockChainParameters in core, since it would require
DB-migrations (I'm pretty sure). Even so, the parameters /are/
jormungandr-specific, and should not belong in core.

- 79f53435242b07a2f877a3af1c63b6e290fbe05d
  Move to separate module + add Ratio type
  
- 2e2054344a09d22848f95ba9bea10942f0cc3cf5
  add rankingEpochConstants function
  
- b905f2cc028b852857e45beef023c0ca03ad5427
  simplify and fix Jormungandr reward params extraction
  
- 3047095c57765b109979e6f4a67945c800f2d178
  finish wiring up incentive parameters with jormungandr's startup
  
- e15f53a45d8bb0489a08ef7f91fe9021a76055af
  add simple unit test to compare ITN params with calculation
  
- 1c2fcb154b3df043f7ea218e35114e3839fe9cb8
  fixup: typo in Ranking module
  
- 99dd776d5a99f8c066b59898ff2bed4f7353708f
  re-generate nix machinery
  
- 8ec5ef6586f02da59d0dd5e7d408b255c57491b4
  additional test vectors for rewards calculation
  
- 905cb71807adc658d33bad90c991da42961e79ef
  make sure pools are sorted by apparent performance in scenario
  
- 2037c6049f72b2274f38d91915d310395bc9c690
  use total block production instead of latest one
    The rationale is that it is quite confusing as many pools
  have only a few blocks produced per epoch (1 or 2) and sometimes
  even 0 for several epochs.
  This makes it even harder to assess whether a pool is doing okay
  or not. Looking at the total number of blocks produced makes more
  sense.

- 22fff45281556d3996f4d086cf7c0c2aa587439d
  add simple properties to test readTotalProduction

# Comments

- I am trying to keep the reward-calculation jörmungandr-specific. I don't
want to extend BlockChainParameters in core, since it would require
DB-migrations (I'm pretty sure). Even so, the parameters /are/
jormungandr-specific, and should not belong in core.

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
[WIP] Calculate rewards from block0 params

I am trying to keep the reward-calculation jörmungandr-specific. I don't
want to extend BlockChainParameters in core, since it would require
DB-migrations (I'm pretty sure). Even so, the parameters /are/
jormungandr-specific, and should not belong in core.

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -1053,3 +1155,91 @@ poolRegistrationsFromBlock (Block _hdr fragments) = do
+
+rewardParamsFromBlock0
+    :: Block
+    -> Maybe (RewardParams, TreasuryTax, Maybe (Word64, Word64))
```
I think I should have introduced a `Ratio` type instead of `(Word64, Word64)`.

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -824,6 +839,63 @@ getConfigParam = label "getConfigParam" $ do
+            2 -> HalvingRewards <$> getHalvingParams
+            _ -> fail $ "Unknown reward parameters with tag " ++ show tag
+
+    -- NOTE: getHalvingParams and getLinearParams are currently identical.
```
Actually, maybe we should leverage that they are identical. Jormungandr does not, but if jörmungandr differentiates them in the future… that's a problem for… the future.

KtorZ replied:
When do you mean, "they" are identical ? What is identical? The shape of the serialized data?

Anviking replied:
Yes.

Anviking committed:
Move to separate module + add Ratio type

Anviking committed:
add rankingEpochConstants function

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -824,6 +836,49 @@ getConfigParam = label "getConfigParam" $ do
+
+
+    getRewardLimitByAbsoluteStake :: Get Ratio
+    getRewardLimitByAbsoluteStake = getRatio
```
This function seems superfluous and adds an unnecessary indirection. I'd suggest to simply use `getRatio`.

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Rewards.hs:
```diff
@@ -0,0 +1,127 @@
+rewardParamsFromBlock0 block = do
+        rewardParams <- find $ \case
+                ConfigRewardFormula x -> Just x
+                _ -> error "missing config reward"
```
Why use `find` and `Maybe` for safety while using `error` for missing values?

Anviking replied:
Oops, this was while debugging, to have more information than `Maybe`. Ultimately an `Either` would be good, I think, but I am not prioritising that.

KtorZ committed:
simplify and fix Jormungandr reward params extraction

KtorZ committed:
finish wiring up incentive parameters with jormungandr's startup

KtorZ committed:
add simple unit test to compare ITN params with calculation

KtorZ committed:
fixup: typo in Ranking module

KtorZ committed:
re-generate nix machinery

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:
```diff
@@ -282,23 +282,19 @@ serveWallet Tracers{..} sTolerance databaseDir hostPref listen backend beforeMai
-    stakePoolLayer nl db metadataDir = do
+    stakePoolLayer nl db block0 metadataDir = do
         void $ forkFinally (monitorStakePools tr nl' db) onExit
+        getEpCst <- maybe (throwIO ErrStartupMissingIncentiveParameters) pure $
```
👍

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Rewards.hs:
```diff
@@ -0,0 +1,169 @@
+    linearAbsorption RewardParams{rFixed,rRatio,rEpochStart,rEpochRate} =
+        if a > c then 0 else c - a
+      where
+        a = r * (n / e)
```
I think there should be a `floor` here.

See: https://github.com/input-output-hk/chain-libs/blob/149e01cf61a9eef2660a6944fe0acc73d490e70f/chain-impl-mockchain/doc/incentives.md#L105-L117

example with rate > 1

Anviking replied:
_Looking more into other things first_

KtorZ replied:
You are pointing to the linear formula but the example from your link is about the halving formula ?

KtorZ replied:
Yet, these are nice examples to use as test vectors perhaps.

KtorZ committed:
additional test vectors for rewards calculation

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -1044,12 +1082,57 @@ overrideFeePolicy linearFee@(LinearFee a b _) override =
+        , desiredNumberOfPools =
+            unsafeMkPositive $ fromIntegral $ maxParticipation poolCapping
+        , totalRewards =
+            rewardsAt (rewardLimit, totalStake) treasuryTax ep rewardFormula
```
super-nitpick: if the name is rewards**At**, the `epoch` should be the first argument, I think.

KtorZ committed:
make sure pools are sorted by apparent performance in scenario

KtorZ committed:
use total block production instead of latest one

  The rationale is that it is quite confusing as many pools
  have only a few blocks produced per epoch (1 or 2) and sometimes
  even 0 for several epochs.
  This makes it even harder to assess whether a pool is doing okay
  or not. Looking at the total number of blocks produced makes more
  sense.

KtorZ committed:
add simple properties to test readTotalProduction

Anviking commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -843,3 +848,10 @@ joinStakePoolWithFixtureWallet ctx = do
+    => (a, f [ApiStakePool])
+    -> (a, f [ApiStakePool])
+sortByPerformance =
+    second (fmap (sortOn (Down . view #apparentPerformance)))
```
This should not be needed!

The problem is that the total rewards are 0 in the integration tests.

If `R` is not 0, the tests pass without `sortByPerformance`.

This highlights an interesting problem: what if the rewards *are* 0 in production?
